### PR TITLE
Feat(mllib): add mlserver mllib implementation

### PR DESCRIPTION
**Pull request by FogDong**
We do not support models that trained by LDA, PowerIterationClustering, StreamingKMeans, FPGrowth in MLlib right now since they do not have the predict func.

If the code and the format is ready to go I'll add docs like README, PTAL /cc @adriangonz

### CODE REVIEW

**Comment by seldondev** [comment]
Hi @FogDong. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/SeldonIO/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2FMLServer).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2FMLServer).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/MLServer/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: custom/mlservermllib/__init__.py
Comment (adriangonz) [inline custom/mlservermllib/__init__.py:6]: In here, probably we only need to export the `MLLibModel` class

File: custom/mlservermllib/models/__init__.py
Comment (adriangonz) [inline custom/mlservermllib/models/__init__.py:3]: The `models` package can probably act as the top-level package for `mlserver_mllib`

File: custom/mlservermllib/models/errors.py
Comment (adriangonz) [inline custom/mlservermllib/models/errors.py:1]: Should we import this class directly from the `mlserver` package instead?

File: custom/mlservermllib/models/mllib.py
Comment (adriangonz) [inline custom/mlservermllib/models/mllib.py:10]: Do these needs to be configurable? You know way more about Spark than I do, so please feel free to correct me if I'm wrong.

File: custom/mlservermllib/models/utils.py
Comment (adriangonz) [inline custom/mlservermllib/models/utils.py:18]: Would it make sense to move this dict into a "const" (as in a "fake const", like `MODEL_DICT`)?

File: custom/setup.cfg
Comment (adriangonz) [inline custom/setup.cfg:1]: This is a good point. Do we need a separate `setup.cfg`. I guess it's not possible to re-use the values from the top-level one.

File: custom/setup.py
Comment (adriangonz) [inline custom/setup.py:25]: Since this package only holds the `mlserver-mllib` extension, we can probably install it directly as a direct dependency rather than an extra. That way, users can just do:

```
pip install mlserver mlserver-mllib
```

File: custom/setup.py
Comment (adriangonz) [inline custom/setup.py:44]: I can see that these `requires` are the same as the top-level ones from the `mlserver` package. Would it make sense to replace them and simply have the following?

```
install_requires=[
	mlserver,
	pyspark
]
```

File: custom/setup.py
Comment (adriangonz) [inline custom/setup.py:52]: Since this is a library, do we need an entry point for this one?

File: custom/setup.py
Comment (adriangonz) [inline custom/setup.py:8]: I was thinking that `mlserver-mllib` could be nicer but I realise that Python doesn't like hyphens much. Although that seems to be convention for "extension packages". For example, if you search for `pytest` plugins, you can see that they are usually named as `pytest-*`: https://pypi.org/search/?q=pytest-

In the end, it probably boils down to preference. Which one would you go for @FogDong ?

**Review by adriangonz — COMMENTED** [review]
This looks amazing @FogDong ! Awesome work!

I've added a couple comments below. Besides those, do you think maybe we should change the `custom/` folder name for something like maybe `runtimes/`? Mainly to differentiate these extra runtimes from the custom use case (as the idea is to support these extensions as well).

So maybe we could have a folder structure like:

```
./runtimes:
	./runtimes/mllib:
		./runtimes/mllib/setup.py
		./runtimes/mllib/mllib.py
		./runtimes/mllib/helpers.py
		...
```

That would give room for other extensions as well. We could even think of moving the SKLearn and XGBoost runtimes under there. The idea is that then people can do the following to pick the subset of runtimes that they want:

```
pip install mlserver mlserver-mllib mlserver-sklearn
```

This is just a very high-level idea though. What are your thoughts on the folder structure @FogDong ?

File: custom/setup.py
Comment (FogDong) [reply custom/setup.py:25]: I'll add this in the README doc.

File: custom/setup.py
Comment (FogDong) [reply custom/setup.py:44]: Put the `mlserver` in the requires means we don't need to `pip install mlserver mlserver-mllib` but just `pip install mlserver-mllib`.

File: custom/setup.py
Comment (FogDong) [reply custom/setup.py:52]: I think we can reuse the mlserver entrypoint

File: custom/mlservermllib/models/mllib.py
Comment (FogDong) [reply custom/mlservermllib/models/mllib.py:10]: The spark conf can be set to local or other spark clusters, ref https://spark.apache.org/docs/latest/submitting-applications.html#master-urls

Since we use mlserver in k8s container, I only add localhost url here right now, but we could make it more  configurable.

Comment (adriangonz) [reply custom/mlservermllib/models/mllib.py]: That makes sense. Let's leave it to localhost for now. We can probably add it to the model parameters if we need to.

File: custom/setup.py
Comment (adriangonz) [reply custom/setup.py]: True. I think that's fair.

**Review by adriangonz — APPROVED** [review]
This looks great now @FogDong! Thanks for going over my comments. I think this is ready to land now. :+1:

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[adriangonz](https://github.com/SeldonIO/MLServer/pull/52#pullrequestreview-521662961 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2FMLServer).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/MLServer/blob/master/OWNERS)~~ [adriangonz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
@FogDong: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 06265d18e77bfedfc7a99d850e001b8d7841913b | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by adriangonz** [comment]
@FogDong the linting pipeline seems to have failed. Could you try to run `make lint` locally? Looking at the build logs, the formatter says that would change the following:

```
# Check if something has changed after generation
git \
        --no-pager diff \
        --exit-code \
        .
diff --git a/runtimes/mllib/mlserver-mllib/mllib.py b/runtimes/mllib/mlserver-mllib/mllib.py
index 13f7ce1..4d92566 100644
--- a/runtimes/mllib/mlserver-mllib/mllib.py
+++ b/runtimes/mllib/mlserver-mllib/mllib.py
@@ -5,9 +5,8 @@ from pyspark import SparkContext, SparkConf


 class MLlibModel(MLModel):
-
     async def load(self) -> bool:
-        conf = SparkConf().set('spark.driver.host', '127.0.0.1')
+        conf = SparkConf().set("spark.driver.host", "127.0.0.1")
         sc = SparkContext(appName="MLlibModel", conf=conf)

         model_uri = self._settings.parameters.uri
diff --git a/runtimes/mllib/mlserver-mllib/utils.py b/runtimes/mllib/mlserver-mllib/utils.py
index 5b912ee..9096b5b 100644
--- a/runtimes/mllib/mlserver-mllib/utils.py
+++ b/runtimes/mllib/mlserver-mllib/utils.py
@@ -1,9 +1,22 @@
 from mlserver import ModelSettings
 from .errors import InvalidMLlibFormat
-from pyspark.mllib.classification import LogisticRegressionModel, NaiveBayesModel, SVMModel
+from pyspark.mllib.classification import (
+    LogisticRegressionModel,
+    NaiveBayesModel,
+    SVMModel,
+)
 from pyspark.mllib.clustering import KMeansModel
-from pyspark.mllib.regression import IsotonicRegressionModel, LinearRegressionModel, LassoModel, RidgeRegressionModel
-from pyspark.mllib.tree import DecisionTreeModel, GradientBoostedTreesModel, RandomForestModel
+from pyspark.mllib.regression import (
+    IsotonicRegressionModel,
+    LinearRegressionModel,
+    LassoModel,
+    RidgeRegressionModel,
+)
+from pyspark.mllib.tree import (
+    DecisionTreeModel,
+    GradientBoostedTreesModel,
+    RandomForestModel,
+)

 MODEL_DICT = {
     "LogisticRegression": LogisticRegressionModel.load,
@@ -16,7 +29,7 @@ MODEL_DICT = {
     "RidgeRegression": RidgeRegressionModel.load,
     "DecisionTree": DecisionTreeModel.load,
     "GradientBoostedTrees": GradientBoostedTreesModel.load,
-    "RandomForest": RandomForestModel.load
+    "RandomForest": RandomForestModel.load,
 }

```

**Comment by adriangonz** [comment]
@FogDong if you run `make fmt`, that should also make sure that the format matches what the linter expects

**Comment by seldondev** [comment]
@FogDong: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 30ff32dec8affd93821fd757a973488e07ea81e2 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**
